### PR TITLE
Fix destination issue, add multiple args and concurrent execution to the `put` command

### PR DESCRIPTION
**Pull request by grantseltzer**
This adds the ability to specify multiple arguments to the `put` command which are then executed using the `go` mechanism. 

On top of this, it fixes an issue that I brought up in #52 in which the 'destination' could be specified but wasn't working the way the description described it. Under my proposed changes the user can specify any number of arguments to upload. If a destination isn't specified (using the `--destination` flag) then the files will be uploaded to the root of the users dropbox directory. 

The issue that is still being posed that may want to be addressed before this is merged is that having many arguments cause 'too many upload attempt' errors to occur. Not sure if this should be handled at this level or at the sdk level. Regardless I'd like to resolve it.

Signed-off-by: grantseltzer grantseltzer@gmail.com

### CODE REVIEW

**Comment by paultopia** [comment]
Just poking my head in here from the issue #52 --- does this also resolve the behavior mentioned in that threat where if the target is a directory, rather than saving uploaded file to that directory, it overwrites that directory altogether with the file?  If not, should that be a separate issue/PR?  (This seems like a particularly dangerous gotcha. I'm imagining hundreds of people entering commands like `dbxcli put logfile.txt some/critical/directory` and then watching their data disappear...

**Comment by GrantSeltzer** [comment]
That's a very good call out, let me do some manual testing and confirm it
works safely. If a fix is needed it definitely should be part of this PR.

File: cmd/put.go [hunk 1]
```diff
@@ -60,68 +62,96 @@ func uploadChunked(dbx files.Client, r io.Reader, commitInfo *files.CommitInfo,
...
-	// The Dropbox API only accepts timestamps in UTC with second precision.
-	commitInfo.ClientModified = time.Now().UTC().Round(time.Second)
-
-	dbx := files.New(config)
-	if contentsInfo.Size() > chunkSize {
-		return uploadChunked(dbx, progressbar, commitInfo, contentsInfo.Size())
+	var waitGroup sync.WaitGroup
+	for _, arg := range args {
+		waitGroup.Add(1)
+		go func(arg string) error {
```
Comment (diwakergupta) [inline cmd/put.go]: Consider defining a helper that takes a waitGroup as arg. Will reduce indent here and make it easier to understand what the high-level flow is.

**Comment by GrantSeltzer** [comment]
@paultopia Thanks again for the call out, files won't be overwriten, but the error wasn't being handled properly (but is now).

@diwakergupta Whenever you get a chance to review :+1:

File: cmd/put.go [hunk 2]
```diff
@@ -60,68 +62,96 @@ func uploadChunked(dbx files.Client, r io.Reader, commitInfo *files.CommitInfo,
...
 var putCmd = &cobra.Command{
-	Use:   "put [flags] <source> [<target>]",
+	Use:   "put [flags] <source>",
 	Short: "Upload files",
 	RunE:  put,
 }
 
 func init() {
 	RootCmd.AddCommand(putCmd)
+	putCmd.Flags().StringP("destination", "d", "", "specify a destination")
```
Comment (diwakergupta) [inline cmd/put.go]: What's the advantage of making this a flag vs. a positional arg? I prefer the latter as it's more intuitive. Most other similar tools (e.g. S3 command line tools, FTP tools) all have a similar behavior.

Comment (GrantSeltzer) [inline cmd/put.go]: My concern was that if you specify multiple arguments, how would we parse what the destination is? If we want to have the files uploaded to the root it wouldn't make sense to have the last argument be the destination. 

I'm not familiar with the S3 command line tools, what does the cli syntax look like?

File: cmd/put.go (hunk 1 above)
Comment (GrantSeltzer) [inline cmd/put.go]: Sure

File: cmd/put.go (hunk 2 above)
Comment (GrantSeltzer) [inline cmd/put.go]: I suppose we can stat the last argument to see if it's an existing directory already, and put it there if it is, and fail if not? The issue then arises that how would we design it if the user wants to upload the files to the root (i.e. no particular directory target)

Comment (GrantSeltzer) [inline cmd/put.go]: @diwakergupta I suppose the simplest solution would be that files will be uploaded to the root unless a path is specified at the end, where the path would have to start with a `/`

That would of course avoid the issue I described above and we wouldn't have to make a decision as for how to handle when to fail or not.

**Comment by GrantSeltzer** [comment]
@diwakergupta Do you have any feedback for my most recent comments?

**Comment by GrantSeltzer** [comment]
@diwakergupta ping

**Comment by diwakergupta** [comment]
@GrantSeltzer sorry -- other priorities meant dbxcli was ignored past many weeks! Let me page stuff back in, will review this PR by tomorrow. Thanks for your patience (and contributions)!

**Comment by GrantSeltzer** [comment]
@diwakergupta No apology needed, I'm in the middle of finals week so I understand having other priorities! (Also means no rush on this).

**Comment by GrantSeltzer** [comment]
@diwakergupta I'd love for development of this to pick back up. Can you take a look at the changes on this?

**Comment by connor-baer** [comment]
Ping :)

**Comment by GrantSeltzer** [comment]
@diwakergupta ping

**Comment by AlJohri** [comment]
@diwakergupta ping

**Comment by Still34** [comment]
Can we get an update on this ancient issue/PR?

**Comment by mbojan** [comment]
Almost two years and counting... Any updates?

**Comment by diwakergupta** [comment]
Hey folks I'm no longer with Dropbox (been 1+ yrs) and I'm not sure what (if anything) Dropbox plans to do with this tool. Sadly maintaining this on my own time has become harder since.

@KarandeepJohar not sure if you're the right person, but perhaps you can tag the right folks.

**Comment by mbojan** [comment]
I'd rather see it closed by a merge @grantseltzer ;) Why?

**Comment by mbojan** [comment]
Too bad :(

**Pull request closed without merging**